### DISCUSSION TITLE
[Feature Request] Add Support for TABLESPACE Specification in Table Definitions for Oracle

### DISCUSSION THREAD

**Discussion by Ineffable22** [sqlalchemy/sqlalchemy#12012] [Ideas]
**Feature Request: Add Support for `TABLESPACE` Specification in Table Definitions for Oracle**

**Summary**

I would like to propose adding support for specifying the `TABLESPACE` in table definitions in SQLAlchemy, specifically for Oracle. This feature would be particularly useful for Oracle users who need to specify the tablespace where the table data will be stored.

**Motivation**

In Oracle databases, specifying the `TABLESPACE` is a common requirement for managing storage and optimizing performance. Currently, SQLAlchemy does not provide a direct way to specify the `TABLESPACE` in table definitions. Adding this feature would enhance SQLAlchemy's support for Oracle and provide more flexibility for users.

**Proposed Solution**

1. **Update `construct_arguments` in `OracleDialect`**:
    - Add the `tablespace` argument to the `construct_arguments` list for the `Table` class.

    ```python
    construct_arguments = [
        (
            sa_schema.Table,
            # old
            {"resolve_synonyms": False, "on_commit": None, "compress": False},
            # new
            {"resolve_synonyms": False, "on_commit": None, "compress": False, "tablespace": None},
        ),
        (sa_schema.Index, {"bitmap": False, "compress": False}),
    ]
    ```
    **Path**: `lib/sqlalchemy/dialects/oracle/base.py`

2. **Modify [`OracleDDLCompiler`](command:_github.copilot.openSymbolFromReferences?%5B%22%22%2C%5B%7B%22uri%22%3A%7B%22scheme%22%3A%22file%22%2C%22authority%22%3A%22%22%2C%22path%22%3A%22%2Fc%3A%2FUsers%2Fmigue%2FDesktop%2Ftestingsql%2Freadme.md%22%2C%22query%22%3A%22%22%2C%22fragment%22%3A%22%22%7D%2C%22pos%22%3A%7B%22line%22%3A25%2C%22character%22%3A12%7D%7D%5D%2C%22d21d1a03-d27d-41a3-b1c4-bbc8b71c5a4a%22%5D "Go to definition") to Include [`TABLESPACE`](command:_github.copilot.openSymbolFromReferences?%5B%22%22%2C%5B%7B%22uri%22%3A%7B%22scheme%22%3A%22file%22%2C%22authority%22%3A%22%22%2C%22path%22%3A%22%2Fc%3A%2FUsers%2Fmigue%2FDesktop%2Ftestingsql%2Freadme.md%22%2C%22query%22%3A%22%22%2C%22fragment%22%3A%22%22%7D%2C%22pos%22%3A%7B%22line%22%3A4%2C%22character%22%3A41%7D%7D%5D%2C%22d21d1a03-d27d-41a3-b1c4-bbc8b71c5a4a%22%5D "Go to definition") in [`post_create_table`](command:_github.copilot.openSymbolFromReferences?%5B%22%22%2C%5B%7B%22uri%22%3A%7B%22scheme%22%3A%22file%22%2C%22authority%22%3A%22%22%2C%22path%22%3A%22%2Fc%3A%2FUsers%2Fmigue%2FDesktop%2Ftestingsql%2Freadme.md%22%2C%22query%22%3A%22%22%2C%22fragment%22%3A%22%22%7D%2C%22pos%22%3A%7B%22line%22%3A25%2C%22character%22%3A55%7D%7D%5D%2C%22d21d1a03-d27d-41a3-b1c4-bbc8b71c5a4a%22%5D "Go to definition")**:
    - Include the [`TABLESPACE`](command:_github.copilot.openSymbolFromReferences?%5B%22%22%2C%5B%7B%22uri%22%3A%7B%22scheme%22%3A%22file%22%2C%22authority%22%3A%22%22%2C%22path%22%3A%22%2Fc%3A%2FUsers%2Fmigue%2FDesktop%2Ftestingsql%2Freadme.md%22%2C%22query%22%3A%22%22%2C%22fragment%22%3A%22%22%7D%2C%22pos%22%3A%7B%22line%22%3A4%2C%22character%22%3A41%7D%7D%5D%2C%22d21d1a03-d27d-41a3-b1c4-bbc8b71c5a4a%22%5D "Go to definition") clause at the end of the `CREATE TABLE` statement if the [`tablespace`](command:_github.copilot.openSymbolFromReferences?%5B%22%22%2C%5B%7B%22uri%22%3A%7B%22scheme%22%3A%22file%22%2C%22authority%22%3A%22%22%2C%22path%22%3A%22%2Fc%3A%2FUsers%2Fmigue%2FDesktop%2Ftestingsql%2Freadme.md%22%2C%22query%22%3A%22%22%2C%22fragment%22%3A%22%22%7D%2C%22pos%22%3A%7B%22line%22%3A4%2C%22character%22%3A191%7D%7D%5D%2C%22d21d1a03-d27d-41a3-b1c4-bbc8b71c5a4a%22%5D "Go to definition") option is provided.

    ```python
    if opts["tablespace"]:
        tablespace_name = opts["tablespace"]
        table_opts.append(
            "\n TABLESPACE %s" % self.preparer.quote(tablespace_name)
        )
    ```
    **Path**: `lib/sqlalchemy/dialects/oracle/base.py`

3. **Add [`tablespace`](command:_github.copilot.openSymbolFromReferences?%5B%22%22%2C%5B%7B%22uri%22%3A%7B%22scheme%22%3A%22file%22%2C%22authority%22%3A%22%22%2C%22path%22%3A%22%2Fc%3A%2FUsers%2Fmigue%2FDesktop%2Ftestingsql%2Freadme.md%22%2C%22query%22%3A%22%22%2C%22fragment%22%3A%22%22%7D%2C%22pos%22%3A%7B%22line%22%3A4%2C%22character%22%3A191%7D%7D%5D%2C%22d21d1a03-d27d-41a3-b1c4-bbc8b71c5a4a%22%5D "Go to definition") Argument to the [`Table`](command:_github.copilot.openSymbolFromReferences?%5B%22%22%2C%5B%7B%22uri%22%3A%7B%22scheme%22%3A%22file%22%2C%22authority%22%3A%22%22%2C%22path%22%3A%22%2Fc%3A%2FUsers%2Fmigue%2FDesktop%2Ftestingsql%2Freadme.md%22%2C%22query%22%3A%22%22%2C%22fragment%22%3A%22%22%7D%2C%22pos%22%3A%7B%22line%22%3A8%2C%22character%22%3A129%7D%7D%5D%2C%22d21d1a03-d27d-41a3-b1c4-bbc8b71c5a4a%22%5D "Go to definition") Class**:
    - Add a new [`tablespace`](command:_github.copilot.openSymbolFromReferences?%5B%22%22%2C%5B%7B%22uri%22%3A%7B%22scheme%22%3A%22file%22%2C%22authority%22%3A%22%22%2C%22path%22%3A%22%2Fc%3A%2FUsers%2Fmigue%2FDesktop%2Ftestingsql%2Freadme.md%22%2C%22query%22%3A%22%22%2C%22fragment%22%3A%22%22%7D%2C%22pos%22%3A%7B%22line%22%3A4%2C%22character%22%3A191%7D%7D%5D%2C%22d21d1a03-d27d-41a3-b1c4-bbc8b71c5a4a%22%5D "Go to definition") argument to the [`Table`](command:_github.copilot.openSymbolFromReferences?%5B%22%22%2C%5B%7B%22uri%22%3A%7B%22scheme%22%3A%22file%22%2C%22authority%22%3A%22%22%2C%22path%22%3A%22%2Fc%3A%2FUsers%2Fmigue%2FDesktop%2Ftestingsql%2Freadme.md%22%2C%22query%22%3A%22%22%2C%22fragment%22%3A%22%22%7D%2C%22pos%22%3A%7B%22line%22%3A8%2C%22character%22%3A129%7D%7D%5D%2C%22d21d1a03-d27d-41a3-b1c4-bbc8b71c5a4a%22%5D "Go to definition") class to allow specifying the tablespace in the table definition.

4. **Update Documentation**:
    - Update the documentation to reflect the new feature and provide usage examples.

**Usage Example**

```python
from sqlalchemy import create_engine, Table, Column, Integer, String, MetaData

# Create the engine
engine = create_engine('oracle+cx_oracle://user:password@dsn')

# Define the metadata
metadata = MetaData()

# Define the table with a tablespace
users = Table('users', metadata,
    Column('id', Integer, primary_key=True),
    Column('name', String),
    Column('email', String, unique=True),
    oracle_tablespace='my_tablespace'  # New tablespace argument optional
)

# Create the table in the database
metadata.create_all(engine)

**Comment by CaselIT** [comment]
Hi,

I guess this should be an issue. creating one and closing this discussion. Please follow it there
